src/group_theory/perm/list.lean

```diff
@@ -0,0 +1,417 @@
+/-
+Copyright (c) 2021 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+
+import data.list.rotate
+import group_theory.perm.support
+
+/-!
+# Permutations from a list
+
+A list `l : list α` can be interpreted as a `equiv.perm α` where each element in the list
+is permuted to the next one, defined as `form_perm`. When we have that `nodup l`,
+we prove that `equiv.perm.support (form_perm l) = l.to_finset`, and that
+`form_perm l` is rotationally invariant, in `form_perm_rotate`.
+
+-/
+
+namespace list
+
+variables {α β : Type*}
+
+section form_perm
+
+variables [decidable_eq α] (l : list α)
+
+open equiv equiv.perm
+
+/--
+A list `l : list α` can be interpreted as a `equiv.perm α` where each element in the list
+is permuted to the next one, defined as `form_perm`. When we have that `nodup l`,
+we prove that `equiv.perm.support (form_perm l) = l.to_finset`, and that
+`form_perm l` is rotationally invariant, in `form_perm_rotate`.
+-/
+def form_perm : equiv.perm α :=
+(zip_with equiv.swap l (l.rotate 1)).tail.prod
```

Comment:
Would it be worth the effort to prove (correctly state?) something along the following lines, and take #7504 as a dependency:
```lean
lemma form_perm_apply_eq_iff (xs : list α) (x y : α) :
  form_perm xs x = y ↔ (x \notin xs ∧ x = y) ∨ (∃ h : x ∈ xs, y = next xs x h) :=
sorry
```
(I believe this is correct without needing `nodup xs`, given `form_perm_apply_head` and `form_perm_not_mem_apply`.)

I feel like we have to rely a lot on case-distinction/induction on `xs` when a shorter specification is possible.

Reply:
This comment has left me a bit dead-locked on what to do about this PR and #7504 (which is why I've taken so long). This definition of `form_perm` makes the following evaluation occur:
```
#eval form_perm [1, 2, 3, 2] 2 -- 1
```
So the definition of `next` as it is right now (in the other PR) matches precisely what `form_perm` does (ditto `prev` for the inverse permutation). Do you think that `next` should be changed to not do this?

Reply:
At least in `next`'s current form, I don't think `form_perm_apply_eq_iff` is true:
```
#eval form_perm [0, 0, 1] 1 -- 1
#eval next [0, 0, 1] 1 dec_trivial -- 0
```
I can try to prove it by assuming `nodup xs`

Reply:
I think I found a way to make the proofs easier, without needing extra lemmas. My proposal is to change the definition of `form_perm` to `(zip_with equiv.swap l l.tail).prod` instead. In other words, we move the tail inside the `zip_with`. (More precisely, we replace `(zip_with equiv.swap l (l.rotate 1)).tail` with `(zip_with equiv.swap l (l.rotate 1)).init`, and this is equal to `zip_with equiv.swap l.init l.tail = zip_with equiv.swap l l.tail`.)

The new definition gives the same results if `nodup l`, since you can prove with induction that:
```lean
old_form_perm (x :: y :: z :: xs) = swap y z * swap x z * (zip_with swap (z :: xs) xs).prod
new_form_perm (x :: y :: z :: xs) = swap x y * swap y z * (zip_with swap (z :: xs) xs).prod
```
which are indeed equal if `x ≠ y` and `x ≠ z` (or if `x = y = z`)

The advantage of the new definition is that it is much more "local": without the `rotate` there is no need to remember what the head of `l` used to be. Indeed, the first couple of proofs went a lot smoother. I'll try and see if the remainder of the file also improves.

Reply:
For the record, my "proof" that the two definitions coincide is:
```lean
lemma form_perm_eq : Π (l : list α), form_perm l = (zip_with equiv.swap l.tail l).prod
| [] := rfl
| [x] := rfl
| [x, y] := rfl
| (x :: y :: z :: xs) :=
by { rw [form_perm_cons_cons_cons, form_perm_eq (x :: z :: xs), tail_cons, tail_cons,
         zip_with_cons_cons, zip_with_cons_cons, zip_with_cons_cons,
         prod_cons, prod_cons, prod_cons, ← mul_assoc, ← mul_assoc,
         mul_swap_eq_swap_mul, swap_apply_right],
       -- TODO: use `nodup` to show `(swap y z) x = x`
       sorry }
```

Reply:
Here's my new version: 7c92eecb18, turning this file from 400 to 350 lines. Feel free to use whatever you'd like from it!